model/map/src/main/java/org/keycloak/models/map/common/AbstractMapProviderFactory.java

```diff
@@ -108,13 +106,32 @@ public String getId() {
     }
 
     protected MapStorage<V, M> getStorage(KeycloakSession session) {
-        ProviderFactory<MapStorageProvider> storageProviderFactory = getComponentFactory(session.getKeycloakSessionFactory(),
-          MapStorageProvider.class, storageConfigScope, MapStorageSpi.NAME);
+        ProviderFactory<MapStorageProvider> storageProviderFactory = getProviderFactoryOrComponentFactory(session, storageConfigScope);
         final MapStorageProvider factory = storageProviderFactory.create(session);
-
         return factory.getStorage(modelType);
     }
 
+    public static ProviderFactory<MapStorageProvider> getProviderFactoryOrComponentFactory(KeycloakSession session, Scope storageConfigScope) {
+        ProviderFactory<MapStorageProvider> storageProviderFactory;
+        if (storageConfigScope.get("realmId") == null) {
```

Comment:
What is this condition used for?

Reply:
The idea of that condition is to see if the provider is not configured for a realm. 

I found similar checks in the DefaultComponentFactoryProviderFactory, and thought this would be the way to go. I'd be happy for advice on how to do this differently.

Reply:
I agree that this check belongs here, but now it's shape is not clear; most likely will be an attribute in a realm, or component lookup in it.

Could you please introduce a `boolean hasRealmSpecificStorage(session) { return false; }` method and use it instead of the `null` check?

Additional info: The `DefaultComponentFactoryProviderFactory` (or more specifically, `KeycloakModelUtils.getComponentFactory`) is a shortcut to use realm-specific component to configure system-wide provider and does not map directly to the condition here.